src/EditorFeatures/Core/Tagging/TaggerThreadCoordinator.cs

```diff
@@ -0,0 +1,192 @@
+
+            /// <summary>
+            /// Original cancellation token controlling the work the tagger needs to do.
+            /// </summary>
+            private readonly CancellationToken _cancellationToken;
+
+            /// <summary>
+            /// Completion source tracking the overall async lifetime of <see cref="_work"/>, including the time spent
+            /// within the <see cref="TaggerThreadCoordinator"/>.  The coordinator itself does not block on <see
+            /// cref="_work"/> executing, so this lets the owning tagger still know that it is occurring, and that it
+            /// can itself await that work.
+            /// </summary>
+            private readonly TaskCompletionSource<bool> _completionSource;
+
+            /// <summary>
+            /// Registration to hear about if the work it canceled before we get around to executing it.  This allows
+            /// the tagger itself to hear about the cancellation and respond to it immediatel.  Disposed once we finally
+            /// remove this work item from the coordinator's queue.
+            /// </summary>
+            private readonly CancellationTokenRegistration _registration;
+
+            public TaggerWork(Func<CancellationToken, Task> work, CancellationToken cancellationToken)
+            {
+                _work = work;
+                _cancellationToken = cancellationToken;
+                var completionSource = new TaskCompletionSource<bool>();
+                _completionSource = completionSource;
+                _registration = cancellationToken.Register(() => completionSource.TrySetCanceled(cancellationToken));
+            }
+
+            public Task Task => _completionSource.Task;
+
+            /// <summary>
+            /// Go and execute the desired tagger work once the coordinator wakes up and attempts to execute a batch of items.
+            /// </summary>
+            public void PerformWork()
+            {
+                if (_cancellationToken.IsCancellationRequested)
+                {
+                    // Reasonable chance the tagger was canceled between when it enqueued work and when we're running.
+                    // Just clear it out and return.  Fast path that to avoid any actual async/await task work.
+                    _completionSource.TrySetCanceled(_cancellationToken);
+                    _registration.Dispose();
+                }
+                else
+                {
+                    // Otherwise, kick off the work and have is signal to the originator (through the completion source)
+                    // when it is finished.  We do this in a fire-and-forget fashion.  The async tagging work the queue
+                    // is doing shouldn't impact the ability for the threading coordinator to move to the next batch of
+                    // work (once the tagging work moves to the bg).
+                    _ = PerformWorkAsync();
```

Comment:
📝 This needs to be tracked with a listener

Reply:
i'll doc, but it actually doesn't.  That's because the tagger itself has listening already tracked in its work queue, and those work queues are given the *task* representing this work in AddUIWorkAsync below.  So there's no awareness at the higher level that hte underlying way that high-level task is accomplished is with a queue+fire-and-forget.  Subtle, but should be correct.

Reply:
doc'ed

Reply:
I did catch that behavior here, but opted to include a token because there is no way to statically enforce correctness, and tracking down a failure would be extremely difficult. Remember that in production scenarios, there is no overhead for including the listener.

Reply:
okey dokey.  i'll add, with that info.